### PR TITLE
Init

### PR DESCRIPTION
**Pull request by eranp22**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,77 @@
...
+| --- | --- | --- | --- |
+| url | The Logz.io Listener URL for your region with port 8071. https://listener.logz.io:8071 | Required | - |
+| token | Your Logz.io log shipping token securely directs the data to your Logz.io account. | Required | - |
+
+**Jumpcloud**
+| Parameter Name | Description | Required/Optional | Default |
+| --- | --- | --- | --- |
+| jumpcloud_api | A dictionary containing the JumpCloud API configurations. | Required | - |
+| url | The URL for the JumpCloud Insights Directory API endpoint. https://api.jumpcloud.com/insights/directory/v1/events. | Required | - |
+| start_date | The start date and time for querying the JumpCloud API in UTC time with the format of %Y-%m-%dT%H:%M:%S.%fZ. For example: 2023-05-04T12:30:00.000000Z. | Optional | The current date and time. |
```
Comment (ShiranAvidov) [inline README.md:34]: lets move `start_date` to be next to `time_interval`

File: README.md [hunk 2]
```diff
@@ -1,2 +1,77 @@
...
+ token: "<<LOGZIO_TOKEN>>"
+
+jumpcloud_api:
+   url: https://api.jumpcloud.com/insights/directory/v1/events
+   start_date:
+   credentials:
+     token: "<<JUMPCLOUD_API_TOKEN>>"
+   settings:
+     time_interval:
+```
```
Comment (ShiranAvidov) [inline README.md:51]: you need to remove `url`

File: src/logzio_shipper.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    @property
+    def value(self) -> str:
+        return self._value
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+VERSION = "1.0.1"
```
Comment (ShiranAvidov) [inline src/logzio_shipper.py]:
```suggestion
VERSION = "1.0.0"
```

File: src/logzio_shipper.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        self._bulk_size = enriched_log_size
+
+    def send_to_logzio(self) -> None:
+        if self._logs is None:
+            return
+
+        try:
+            headers = {"Content-Type": "application/json",
+                       "Content-Encoding": "gzip",
+                       "Logzio-Shipper": "logzio-azure-blob-trigger/v{0}/0/0.".format(VERSION)}
```
Comment (ShiranAvidov) [inline src/logzio_shipper.py]:
```suggestion
                       "Logzio-Shipper": "logzio-jumpcloud/v{0}/0/0.".format(VERSION)}
```

File: src/logzio_shipper.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                logger.error("The token is missing or not valid. Make sure you’re using the right account token.")
+                raise
+
+            logger.error("Somthing went wrong. response: {}".format(e))
+            raise
+        except Exception as e:
+            logger.error("Something went wrong. response: {}".format(e))
+            raise
+
+    def add_custom_field_to_list(self, custom_field: ApiCustomField) -> None:
```
Comment (ShiranAvidov) [inline src/logzio_shipper.py]: you can remove this method

File: src/manager.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            self.jumpcloud_time_interval = 5
+        return True
+
+
+    def write_last_time_to_file(self,last_time_event:str) -> None:
+        with open(self.last_start_dates_file, 'w') as f:
+            f.write(last_time_event)
+
+    class logzio_error(Exception):
+        pass
```
Comment (ShiranAvidov) [inline src/manager.py:157]: add here new line

File: src/manager.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                logger.error(f"HTTP Error: {errh}")
+        except requests.exceptions.ConnectionError as errc:
+            raise self.jumpcloud_api_error(f"Error Connecting: {errc}")
+        except requests.exceptions.Timeout as errt:
+            raise self.jumpcloud_api_error(f"Timeout Error: {errt}")
+        except requests.exceptions.RequestException as err:
+            raise self.jumpcloud_api_error(f"Something went wrong: {err}")
+
+    def run(self):
+
```
Comment (ShiranAvidov) [inline src/manager.py]: remove the new line

File: src/manager.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+                        event["changes"] = self.changesValueToJSON(event["changes"])
+                        event_str = json.dumps(event)
+                    else:
+                        event_str = json.dumps(event)
+                    self.logzio_shipper.add_log_to_send(event_str)
+
+                self.logzio_shipper.send_to_logzio()
+                self.last_time_event = events[-1]['timestamp']
+            else:
+                logger.info("No events")
```
Comment (ShiranAvidov) [inline src/manager.py:179]: if there are no events, you dont want to update the last_time_event. you want the next iteration to have same last_time_event. you need to end the method.

File: src/manager.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+            self.last_time_event = self.last_time_plus_sec(self.last_time_event)
+            self.write_last_time_to_file(self.last_time_event)
+        except Exception as e:
+           raise self.logzio_error("Failed to send data to Logz.io... {}".format(e))
```
Comment (ShiranAvidov) [inline src/manager.py:191]: add here new line

File: src/manager.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+    class logzio_error(Exception):
+        pass
+    def send_events_to_logzio(self, events) -> None:
+        """
+        Sends a list of events to Logz.io.
+
+        :param events: The list of events to send to Logz.io.
+        """
+        self.logzio_shipper = LogzioShipper(self.logzio_url, self.logzio_token)
```
Comment (ShiranAvidov) [inline src/manager.py]: you need to put this line outside of this method. this method is being called every iteration.

File: src/manager.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+                        event_str = json.dumps(event)
+                    self.logzio_shipper.add_log_to_send(event_str)
+
+                self.logzio_shipper.send_to_logzio()
+                self.last_time_event = events[-1]['timestamp']
+            else:
+                logger.info("No events")
+            self.last_time_event = self.last_time_plus_sec(self.last_time_event)
+            self.write_last_time_to_file(self.last_time_event)
+        except Exception as e:
```
Comment (ShiranAvidov) [inline src/manager.py:190]: you need to catch different exceptions from the logio_shipper. if you get 400, 401 or url error you want to exit the application. if you get 500 or connection error you want to try again in the next iteration.

File: src/manager.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    def run(self):
+
+        signal.signal(signal.SIGINT, self.handle_sigint)
+        if not self.read_config():
+            return
+        while True:
+            try:
+                events = self.request_events_jumpcloud()
+                self.send_events_to_logzio(events)
+            except self.jumpcloud_api_error as e:
```
Comment (ShiranAvidov) [inline src/manager.py:233]: here you need to add another catch for logzio_error

**Review by ShiranAvidov — CHANGES_REQUESTED** [review]
almost done :)

File: src/manager.py (hunk 9 above)
Comment (eranp22) [reply src/manager.py:191]: where on the raise function?

File: src/manager.py (hunk 10 above)
Comment (eranp22) [reply src/manager.py]: So I will do it on the read_config function?

File: src/manager.py (hunk 11 above)
Comment (eranp22) [reply src/manager.py:190]: Can you give me an example I can do it?
For example, if I got error 500 from logz.io.

File: src/manager.py (hunk 10 above)
Comment (eranp22) [reply src/manager.py]: Done

File: src/manager.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+                        event_str = json.dumps(event)
+                    self.logzio_shipper.add_log_to_send(event_str)
+
+                self.logzio_shipper.send_to_logzio()
+                self.last_time_event = events[-1]['timestamp']
+            else:
+                logger.info("No events")
+        except Exception as e:
+           raise self.logzio_error("Failed to send data to Logz.io... {}".format(e))
+        except requests.HTTPError as e:
```
Comment (ShiranAvidov) [inline src/manager.py:180]: this except should be before except Exception as e

File: src/manager.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+                raise self.jumpcloud_api_error("HTTP Error: 400 Invalid/unknown query JSON body")
+            else:
+                logger.error(f"HTTP Error: {errh}")
+        except requests.exceptions.ConnectionError as errc:
+            raise self.jumpcloud_api_error(f"Error Connecting: {errc}")
+        except requests.exceptions.Timeout as errt:
+            raise self.jumpcloud_api_error(f"Timeout Error: {errt}")
+        except requests.exceptions.RequestException as err:
+            raise self.jumpcloud_api_error(f"Something went wrong: {err}")
+
```
Comment (ShiranAvidov) [inline src/manager.py:224]: please add except Exception as e:

`except Exception as e:
      raise self.jumpcloud_api_error(f"Something went wrong: {err}")`

**Review by ShiranAvidov — CHANGES_REQUESTED** [review]

**Review by ShiranAvidov — APPROVED** [review]

**Pull request merged**